### PR TITLE
feat(select): use selection value as a label as fallback in production

### PR DESCRIPTION
**Pull request by ismay**
https://dhis2.atlassian.net/browse/LIBS-401

Currently this doesn't show the value in the dropdown, only in the input. That means it can be programmatically selected, but once the selection is removed the user can't select it again through the dropdown. Made the most sense to me since it isn't part of the options, so this way it fails gracefully in production, but doesn't behave too weirdly.

### CODE REVIEW

**Review by Mohammer5 — APPROVED** [review]
I have a question:

If there's a selected option, I'd expect the app to always provide that.
That way the select always know the label. There are other way of providing a label for a selected label (e.g. through a prop), but allowing this pattern (of not providing the option) feels like it'd encourage apps to not provide it more often than not which will result in an inconsistent UI (sometimes the label is shown, sometimes the value)

Response by @ismay on slack:

> > If there's a selected option, I'd expect the app to always provide that.
> 
> Me too. That's why we initially always threw an error (even in production). With the scheduler app there was an edge case (I believe Jan made a change directly via the api that resulted in a selected value that the app didn't know, causing a crash) and the decision was to not crash in production for this kind of situation. It's still a misconfiguration, but now we no longer crash when it occurs in production.
An argument could still be made against the solution in this PR. I think the one usecase we've found (the scheduler one), is a very weak argument in favor of the change, and I fail to see any other convincing usecase. But, nevertheless, this is the fix if this is the behaviour we want.
>
> > feels like it'd encourage apps to not provide it more often than not which will result in an inconsistent UI (sometimes the label is shown, sometimes the value)
>
> Yeah, it does make it a little easier to do so. Though, it will still throw outside of production, so it's still a thing you pretty much have to resolve. Unless it somehow only occurs in production. In that case this fix will make it harder to catch that, but on the other hand, it will also not immediately crash the app (which is maybe also too severe).

**Review by tomzemp — COMMENTED** [review]
Looks good to me

I'm wondering though because on the JIRA issue you note:
> I think my vote would be for doing the former (“Make it clear to users that the select has an error and needs an action”). We could show an error icon and a clear label that the current option is invalid. 

and here we do not seem to have any indication that `selected` option is invalid?

Also, I know you've added the tests for the issue, but it's a bit funny that we do not have a baseline test for when things go right and `selected` is valid 😛

File: components/select/src/multi-select/__tests__/selection-list.test.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import React from 'react'
+import { MultiSelectOption } from '../../multi-select-option/index.js'
+import { SelectionList } from '../selection-list.js'
+import '@testing-library/jest-dom'
+
+describe('<SelectionList />', () => {
+    const env = process.env
+
+    beforeEach(() => {
+        jest.resetModules()
```
[the remark this answers is not in the document]
Comment (ismay) [reply components/select/src/multi-select/__tests__/selection-list.test.js:11]: Yeah, I took this from a post on the matter. I think this is done more to ensure that modules that have already parsed process.env will reevaluate it after it has been reset.

File: components/select/src/single-select/__tests__/selection.test.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import React from 'react'
+import { SingleSelectOption } from '../../single-select-option/index.js'
+import { Selection } from '../selection.js'
+import '@testing-library/jest-dom'
+
+describe('<Selection />', () => {
+    const env = process.env
+
+    beforeEach(() => {
+        jest.resetModules()
```
[the remark this answers is not in the document]
Comment (ismay) [reply components/select/src/single-select/__tests__/selection.test.js:11]: https://github.com/dhis2/ui/pull/1231#discussion_r1099970782

**Comment by ismay** [comment]
> Looks good to me
> 
> I'm wondering though because on the JIRA issue you note:
> 
> > I think my vote would be for doing the former (“Make it clear to users that the select has an error and needs an action”). We could show an error icon and a clear label that the current option is invalid.
> 
> and here we do not seem to have any indication that `selected` option is invalid?

Ah yeah, we had a discussion on slack where ended up with the solution in this PR. See here: https://dhis2.slack.com/archives/CPGUFVC56/p1675248508384359. So:

- Throw error on non-existent selected option in !production
- Console.error on non-existent selected option in production, and show the value as the label

> Also, I know you've added the tests for the issue, but it's a bit funny that we do not have a baseline test for when things go right and `selected` is valid stuck_out_tongue

Not in jest, but we have a ton of cypress tests for it.

File: components/select/src/multi-select/__tests__/selection-list.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ismay) [reply components/select/src/multi-select/__tests__/selection-list.test.js:12]: I added an explanatory comment. Let me know what you think

File: components/select/src/multi-select/__tests__/selection-list.test.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                    onChange={() => {}}
+                />
+            )
+
+            expect(getByText(selected)).toBeInTheDocument()
+        })
+
+        it('log an error for a selection that is not in the options', () => {
+            process.env.NODE_ENV = 'production'
+            const spy = jest.spyOn(console, 'error')
```
[the remark this answers is not in the document]
Comment (ismay) [reply components/select/src/multi-select/__tests__/selection-list.test.js]: Yeah that's true. I usually tend to want to contain everything that's happening in a test as much as possible in that test, to make it easier to scan, but your suggestion seems easy enough to grasp for someone reading the test. I used `restoreAllMocks`, it fully resets the mocked method.

**Review by tomzemp — APPROVED** [review]

**Pull request merged**